cli/assets/templates/rust/src/alloc.rs

```diff
@@ -0,0 +1,29 @@
+use buddy_alloc::{BuddyAllocParam, FastAllocParam, NonThreadsafeAlloc};
+
+extern "C" {
+    pub(crate) static __heap_base: usize;
+}
+
+/// Returns a pointer with offset `__heap_base`.
+/// based on https://github.com/rustwasm/wee_alloc/issues/88#issuecomment-581404618
+macro_rules! heap_ptr {
+    ($t:ty) => {{
+        &crate::alloc::__heap_base as *const _ as *const () as *const $t
+    }};
+}
+
+pub(crate) use heap_ptr;
+
+// These values can be tuned
+const FAST_HEAP_SIZE: usize = 4 * 1024; // 4 KB
+const HEAP_SIZE: usize = 16 * 1024; // 16 KB
+const LEAF_SIZE: usize = 16;
+
+static mut FAST_HEAP: [u8; FAST_HEAP_SIZE] = [0u8; FAST_HEAP_SIZE];
+
+#[global_allocator]
+static ALLOC: NonThreadsafeAlloc = unsafe {
+    let fast_param = FastAllocParam::new(FAST_HEAP.as_ptr(), FAST_HEAP_SIZE);
+    let buddy_param = BuddyAllocParam::new(heap_ptr!(u8), HEAP_SIZE, LEAF_SIZE);
```

Comment:
Nice!!

Can we use `__heap_base` to dynamically size the heap size instead of needing a `HEAP_SIZE` constant? Since the total memory in WASM-4 is 64 KB, we should be able to calculate the heap size to be `65535 - __heap_base`?

Reply:
I've tried to achieve that using several strategies but even in nightly mode I get an error when I try it.

### Explanation

`__heap_base`  taken from the[`extern` block is **always 0**](https://github.com/rustwasm/wee_alloc/issues/88#issuecomment-581547608) but you could extract its value using a simple cast `as u32` unfortunately
 I've not found a way to write `65535 - __heap_base`.

#### Things I've tried

1.  `BuddyAllocParam::new(heap_ptr!(u8), 65535 - (heap_ptr!(u8) as usize), LEAF_SIZE);`

```
Compiler error:
pointers cannot be cast to integers during const eval.
at compile-time, pointers do not have an integer value
avoiding this restriction via `transmute`, `union`, or raw pointers leads to compile-time undefined behavior
```

2.  `BuddyAllocParam::new([...](heap_ptr!(u8).offset_from(0 as *const u8) as usize), [...])`

Requires switch to nightly and flag `#![feature(const_ptr_offset_from)]`

```
Compiler error:
null pointer is not a valid pointer for this operation
more info here: rustc --explain E0080
```

Essentially rust does not support pointer arithmetic when working with `const` values.

I'm wondering if it would be safer to use two array instead, like the snippet you showed in the other thread.
What do you think?

Reply:
I think the idea about using `__heap_base` was that so we could dynamically size the heap. If that doesn't work, I guess it might be better to just use a statically allocated array.

Here's a thought... do we even need to calculate a valid heap size? What happens if we just set the `HEAP_SIZE` constant to something huge like a megabyte? It will of course actually run out of memory far sooner, but the wasm runtime can catch that, rather than buddy-alloc.

Reply:
> Here's a thought... do we even need to calculate a valid heap size? What happens if we just set the HEAP_SIZE constant to something huge like a megabyte?

I've tried with something far smaller, `65535`, but it does not work.

### Test code

```rust
    let mut vecc: Vec<u32> = vec![1, 2, 3];

    for n in 0..20 {
        trace(format!("{:?}", n));
        vecc.push(n);
    }
```

it crashes as soon you reach the 8th push:
I haven't checked the source code but I assume that it tries to ensure capacity and fails.

Using a static heap instead works and the vector grows much further.
Moreover it is the way officially supported by the [allocator](https://github.com/jjyr/buddy-alloc/blob/master/examples/non_threadsafe_demo.rs).